pkg/controller/core/clusterqueue_controller.go

```diff
@@ -73,6 +77,25 @@ func (r *ClusterQueueReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	ctx = ctrl.LoggerInto(ctx, log)
 	log.V(2).Info("Reconciling ClusterQueue")
 
+	if cqObj.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&cqObj, ClusterQueueFinalizerName) {
+			controllerutil.AddFinalizer(&cqObj, ClusterQueueFinalizerName)
+			if err := r.client.Update(ctx, &cqObj); err != nil {
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(&cqObj, ClusterQueueFinalizerName) {
+			if r.cache.ClusterQueueEmpty(cqObj.Name) {
```

Comment:
We should atomically decommission the CQ at this point (mark it as pending).

Otherwise, in the time there is between we queried `ClusterQueueEmpty` and the time the ClusterQueue is actually removed, other workloads could be admitted by the scheduler.